create-calendar-collection.js

```diff
@@ -0,0 +1,280 @@
+    let firstSundayDate;
+    if (firstSundayDayOffset !== 0) {
+        const dateOffsetFromEndOfPreviousMonth = firstSundayDayOffset + 1;
+        firstSundayDate = new Date(endOfPreviousMonthDate);
+        firstSundayDate.setUTCDate(firstSundayDate.getUTCDate() + dateOffsetFromEndOfPreviousMonth);
+    } else {
+        firstSundayDate = new Date(firstOfMonthDate);
+    }
+    const calendarMonthEndDay = getDaysInMonth(year, monthIndex);
+    const endOfMonthDate = new Date(Date.UTC(year, monthIndex, calendarMonthEndDay, hoursOffsetBetweenLocalAndUTC));
+    const endOfMonthDayName = endOfMonthDate.toLocaleDateString('en-US', {
+        weekday: 'long',
+    }).toLowerCase();
+    const lastSaturdayDayOffset = getCountOfDaysToSaturday(endOfMonthDayName);
+    let lastSaturdayDate;
+    if (lastSaturdayDayOffset !== 0) {
+        const nextMonthIndex = monthIndex < 11 ? monthIndex + 1 : 0;
+        const nextMonthYear = nextMonthIndex > 0 ? year : year + 1;
+        const nextMonthFirstSaturdayDate = lastSaturdayDayOffset;
+        lastSaturdayDate = new Date(Date.UTC(nextMonthYear, nextMonthIndex, nextMonthFirstSaturdayDate, hoursOffsetBetweenLocalAndUTC));
+    } else {
+        lastSaturdayDate = new Date(endOfMonthDate);
+    }
+    let numberOfDaysFromPreviousMonth = previousMonthEndDay - firstSundayDate.getUTCDate() + 1;
+    if (firstSundayDate.getUTCMonth() === monthIndex) {
+        numberOfDaysFromPreviousMonth = 0;
+    }
+    let numberOfDaysFromNextMonth = lastSaturdayDate.getUTCDate();
+    if (lastSaturdayDate.getUTCMonth() === monthIndex) {
+        numberOfDaysFromNextMonth = 0;
+    }
+    const numberOfDaysInCalendar = numberOfDaysFromPreviousMonth + calendarMonthEndDay + numberOfDaysFromNextMonth;
+    const NUMBER_OF_DAYS_IN_WEEK = 7;
+    const numberOfWeeksInCalendar = numberOfDaysInCalendar / NUMBER_OF_DAYS_IN_WEEK;
+    const firstSundayDateUTCMonth = firstSundayDate.getUTCMonth() + 1;
+    const lastSaturdayDateUTCMonth = lastSaturdayDate.getUTCMonth() + 1;
+    const firstSundayUTCDate = firstSundayDate.getUTCDate();
+    const lastSaturdayUTCDate = lastSaturdayDate.getUTCDate();
+    const formattedDateOfFirstSunday = `${firstSundayDate.getUTCFullYear()}-${firstSundayDateUTCMonth < 10 ? `0${firstSundayDateUTCMonth}` : firstSundayDateUTCMonth}-${firstSundayUTCDate < 10 ? `0${firstSundayUTCDate}` : firstSundayUTCDate}`;
+    const formattedDateOfLastSaturday = `${lastSaturdayDate.getUTCFullYear()}-${lastSaturdayDateUTCMonth < 10 ? `0${lastSaturdayDateUTCMonth}` : lastSaturdayDateUTCMonth}-${lastSaturdayUTCDate < 10 ? `0${lastSaturdayUTCDate}` : lastSaturdayUTCDate}`;
+
+    const calendarEntryTpl = `---
+layout: calendar
+year: ${year}
+month_index: ${monthIndex}
+month_number: ${month}
+date_of_first_sunday: '${formattedDateOfFirstSunday}'
+date_of_last_saturday: '${formattedDateOfLastSaturday}'
+calendar_view_week_count: ${numberOfWeeksInCalendar}
```

Comment:
Correct me if i am wrong - I have not gone through all the code yet:

Are we doing the math in js for laying out the calendar and then letting jekyll build a page using that? If so, I would have preferred js to do everything and produce dumb html files for jekyll to enrich. This way, one language and one code builds the calander.

Reply:
This script automates the creation of Calendar Collection entries so that they don't have to be manually created, and are in sync with what needs to be shown in the calendar events pages. Technically speaking, they can be created manually. This is to prevent that. This doesn't actually build the calendar pages. Only the meta data that specifies what is shown. Not how it's shown. That is handled in `_layouts/calendar.html` and is easy for people who are not JavaScript programmers to  maintain in HTML, SCSS with Liquid logic.